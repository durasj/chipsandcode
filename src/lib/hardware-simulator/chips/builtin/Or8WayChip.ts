import type { ChipPin } from '../Chip';
import type Chip from '../Chip';

/**
 * Or8Way chip implemented natively
 */
class Or8WayChip implements Chip {
  public readonly name = 'Or8Way';

  private in: boolean[];
  private out: boolean[];

  constructor(private width = 8) {
    this.in = new Array(width).fill(false);
    this.out = [false];
  }

  setInput(name: 'in', value: boolean[]) {
    this[name] = value;
  }

  getOutput() {
    return this.out;
  }

  getPins(): Map<string, ChipPin> {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['in', { type: 'input', width: this.width, state: this.in, connections: [] }],
      ['out', { type: 'output', width: this.width, state: this.out, connections: [] }],
    ]);
  }

  public run() {
    this.out = [this.in.reduce((acc, p) => acc || p, false)];
  }
}

export default Or8WayChip;
