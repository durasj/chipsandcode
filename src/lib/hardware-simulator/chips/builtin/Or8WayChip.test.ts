import { describe, expect, it } from 'vitest';

import Or8WayChip from './Or8WayChip';

describe('Or8Way Chip', () => {
  it('Provides interface info', async () => {
    const chip = new Or8WayChip();

    expect(chip.name).toBe('Or8Way');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new Or8WayChip();

    const values = [
      [[false, false, false, false, false, false, false, false], [false]],
      [[false, true, false, false, false, false, false, false], [true]],
      [[false, true, false, false, false, false, false, true], [true]],
      [[true, false, false, false, false, false, false, false], [true]],
      [[true, true, true, true, true, true, true, true], [true]],
    ];

    values.forEach(([input, out]) => {
      chip.setInput('in', input);
      chip.run();
      expect(chip.getOutput()).toStrictEqual(out);
    });
  });
});
