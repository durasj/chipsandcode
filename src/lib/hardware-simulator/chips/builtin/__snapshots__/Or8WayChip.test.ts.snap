// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Or8Way Chip > Provides interface info 1`] = `
Map {
  "in" => {
    "connections": [],
    "state": [
      false,
      false,
      false,
      false,
      false,
      false,
      false,
      false,
    ],
    "type": "input",
    "width": 8,
  },
  "out" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 8,
  },
}
`;
